enhancements/microshift/microshift-router-configuration.md

```diff
@@ -0,0 +1,621 @@
+HTTP/1.0 503 Service Unavailable
+pragma: no-cache
+cache-control: private, max-age=0, no-cache, no-store
+content-type: text/html
+```
+
+Depending on the log level configuration of ovnk we might get extra logs in
+`ovnkube-master` pod whenever a host port is opened or closed.
+
+The current LoadBalancer controller should also log whenever there are changes
+to IPs and/or ports.
+
+Since firewalld is not a valid way of dropping traffic to these ports there is
+a need for another way. There are two options:
+- Disable the router.
+- Use NetworkPolicy resources.
+
+Disabling the router might be disruptive, as not all traffic may be coming from
+external sources. This procedure also requires a restart, which might not be
+desirable either.
+
+Using NetworkPolicy resources a user may control the traffic that is allowed
+into the cluster. By creating a deny-all policy the router will still remain
+operative for internal traffic, while rejecting all external traffic. An
+example follows:
+```yaml
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  creationTimestamp: "2023-10-31T12:30:03Z"
+  generation: 1
+  name: block-router
+  namespace: openshift-ingress
+  resourceVersion: "2858"
+  uid: d0f01408-089c-41e4-816b-82b7205fc1db
+spec:
+  ingress:
+  - from:
+    - ipBlock:
+        cidr: 10.42.0.0/16
+  podSelector:
+    matchLabels:
+      ingresscontroller.operator.openshift.io/deployment-ingresscontroller: default
+  policyTypes:
+  - Ingress
+```
+Assuming the pod network is configured to use CIDR `10.42.0.0/16`, the above
+policy will block all traffic not coming from that network.
+
+If a user needs to disable all traffic, external and internal, the following
+network policy will do:
+```yaml
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  creationTimestamp: "2023-10-30T11:51:35Z"
+  generation: 1
+  name: block-router
+  namespace: openshift-ingress
+  resourceVersion: "228099"
+  uid: ae86d503-af73-417a-ac1c-073325344322
+spec:
+  podSelector:
+    matchLabels:
+      ingresscontroller.operator.openshift.io/deployment-ingresscontroller: default
+  policyTypes:
+  - Ingress
+```
+
+Or explicitly allowing certain ports, blocking the rest:
+```yaml
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  creationTimestamp: "2023-10-30T13:48:58Z"
+  generation: 1
+  name: block-router
+  namespace: openshift-ingress
+  resourceVersion: "1891"
+  uid: 93c893d8-9e57-4e9f-9bc7-2eac4eb054f6
+spec:
+  ingress:
+  - ports:
+    - port: 443
+      protocol: TCP
+  podSelector:
+    matchLabels:
+      ingresscontroller.operator.openshift.io/deployment-ingresscontroller: default
+  policyTypes:
+  - Ingress
+```
+This policy will only allow traffic enter through port 443, while blocking 80.
+
+Any of these methods is equivalent to using firewalld to block traffic, but
+using the CNI instead of a host level setting. This provides greater
+flexibility, as it is self contained within MicroShift.
```

Comment:
Are we considering automatically generating these policies based on configuration settings?

Reply:
Yes, that is the idea. Will add more details in the design details section.

Reply:
Why wouldn't we just tell users to set up their own policies?

Reply:
Based on our conversation earlier, I think we can focus in this enhancement on the configuration options needed to let the user control where the router listens. We can document the use of the NetworkPolicy API to control access to that endpoint separately, since there's no code change needed for that.